### PR TITLE
#9JpkBREJ ft-register-drivers-and-operators-with proper validations

### PR DESCRIPTION
**Pull request by Niyiragira**
### **What does this PR do?**
requests to merge ft-register-driversd-operators-9JpkBREJ with develop
### **Description of Task to be completed?**
Created route for registering operators and drivers
Generate randomly passwords for registered users and hash them for security
Validate users
Check if a user is not registered twice
Send them email of the unhashed password
### **How should this be manually tested?**
-Visit this PR deployed app to perform below tasks https://phantom-caba-ft-registe-doghsw.herokuapp.com/api-docs/
-Clone this repo

-git checkout origin/ft-register-drivers-and-operators-9JpkBREJ

-install all packages

-make sure your db is connected with the help of .env.example file
-log in your email using help of .env.example file
**N.B:** two emails are required: the one to send password in auth function and the one to receive it
please use .env.example

-do npm run migrate

-go to your postman or browse http://localhost:5000/api-docs/
POST login using these to have admin permissions :**email:admin@phantom.com** and **password:admin**
POST at http://localhost:5000/api/v1/users/ to send an email to a new user and register them
5.Use one of the users registered password to login
use the token to write into headers  
_"Bearer **token**"_  as shown in below screenshot
### **What are the relevant pivotal tracker stories?**

[9JpkBREJ](https://trello.com/c/9JpkBREJ/10-register-drivers-operators
### **Screenshots (if appropriate)**
Screen Shot 2020-10-15 at 15 32 06
-55f8f1deede2.png">

### CODE REVIEW

**Comment by Nkbtemmy** [comment]
Why your PR doesn't deployed ?????

**Comment by Niyiragira** [comment]
> Why your PR doesn't deployed ?????

I don't get it too. May it be may mistake?

File: src/language/locales/en.json [hunk 1]
```diff
@@ -17,5 +17,18 @@
...
+	"date of birth must be provided": "date of birth must be provided",
+	"Name must not be empty": "Name must not be empty",
+	"name is required": "name is required",
+	"Email must not be empty": "Email must not be empty",
+	"gender should be not more 8 characters": "gender should be not more 8 characters",
+	"gender is required": "gender is required",
+	"gender must not be empty characters": "gender must not be empty characters",
+	"birthdate is required": "birthdate is required",
+	"date of birth must be atleast six characters": "date of birth must be atleast six characters"
+}
```
Comment (Nkbtemmy) [inline src/language/locales/en.json:35]: new line required

File: src/language/locales/rw.json [hunk 2]
```diff
@@ -15,5 +15,13 @@
...
+	"user doesn't exist": "ukoresha app ntabonetse",
+	"role must be one of [normal, driver, operator, admin]": "role must be one of [normal, driver, operator, admin]",
+	"User with the provided email is already registered.": "Umukoresha hamwe na imeri yatanzwe yamaze kwiyandikisha.",
+	"provide valid email": "tanga imeri yemewe",
+	"Name must be atleast four characters": "Izina rigomba kuba byibuze inyuguti enye",
+	"birthdate must be atleast six characters": "itariki y'amavuko igomba kuba byibuze inyuguti esheshatu",
+	"date of birth must be provided": "itariki y'amavuko igomba gutangwa",
+	"date of birth must be atleast six characters": "date of birth must be atleast six characters",
+	"Name must not be empty":"Izina ntirigomba kubamo ubusa"
+}
```
Comment (Nkbtemmy) [inline src/language/locales/rw.json]: Also here

File: src/utils/generateDefaultRole.js [hunk 3]
```diff
@@ -0,0 +1,4 @@
+exports.generateDefault=()=>{
+    let role ="normal"
+ return role
+}
```
Comment (Nkbtemmy) [inline src/utils/generateDefaultRole.js]: new line here required

**Review by Nkbtemmy — CHANGES_REQUESTED** [review]
You have to add new lines where are required

**Review by Nkbtemmy — APPROVED** [review]

File: package.json [hunk 4]
```diff
@@ -17,6 +17,8 @@
   "author": "Niyiragira",
   "license": "ISC",
   "dependencies": {
+    "@hapi/joi": "^17.1.1",
+    "bcrypt": "^5.0.0",
     "bcryptjs": "^2.4.3",
```
Comment (ceelogre) [inline package.json:22]: Why both bcrypt and bycryptjs?

File: src/controller/userControllers.js [hunk 5]
```diff
@@ -54,6 +59,45 @@ class userController {
...
+              const userEmail  = req.body.email;
+              
+            const doesExist = await Model.User.findOne({
+                where: {email:userEmail}
+            });
+              if(doesExist){
+                res.status(409).json({
+                  message:res.__('User with the provided email is already registered.')
+                })
+                return false
```
Comment (ceelogre) [inline src/controller/userControllers.js:86]: Where will this value be received?

File: src/test/testRegister.js [hunk 6]
```diff
@@ -0,0 +1,179 @@
...
+        chai 
+          .request(app)
+          .post("/api/v1/auth/register")
+          .set("Authorization", "Bearer " + userMock.token.admin)
+          .send(registerMock.emptyDate)
+          .end((err,res)=>{
+            expect(res).to.have.status(400);
+            done();
+          })
+      });
```
Comment (ceelogre) [inline src/test/testRegister.js]: You cannot have all these tests checking only for status code (400). Please add more checks, for example, the error messages in the response body.

File: src/utils/generatePassword.js [hunk 7]
```diff
@@ -0,0 +1,10 @@
+exports.generatePassword = () => {
+    const chars = "abcdefghijklmnopqrstuvwxyzABCDEFGHIJKLMNOPQRSTUVWXYZ0123456789!@#$%&?"
+
+    let password = ''
+    for (let i = 0; i < 12; i++) {
+        let index = Math.floor(Math.random() * Math.floor(chars.length))
+        password = password + chars[index]
+    }
+    return password
+}
```
Comment (ceelogre) [inline src/utils/generatePassword.js:10]: It's good you're not relying on an external library to generate a password.

**Review by ceelogre — CHANGES_REQUESTED** [review]
Look at the picture below about the date issue we've discussed on a call.
Capture d’écran 2020-10-21 à 11 20 27

Also, although you've added lots of test cases (which is a good thing), you need to greatly improve your checks/assertions as I've mentioned in the comment.

Great work on password generation, your work is almost ready, it should be merged today.

**Review by ncutixavier — APPROVED** [review]

**Review by ceelogre — APPROVED** [review]

**Pull request merged**